docs/source/events/event_loop.rst

```diff
@@ -0,0 +1,187 @@
+that it plays in napari, and users interested in creating highly customized
+events and actions are advised to gain at least a little familiarity with the
+Qt event loop.
+
+
+Hooking up your own events
+--------------------------
+
+If you're coming from a background of scripting or working with python in an
+interactive console, thinking in terms of the "event loop" can feel a bit
+strange at time.  Often we write code in a very procedural way: "do this ...
+then do that, etc...". With napari and other GUI programs however, usually you
+hook up a bunch of conditions and to callback functions (e.g. "If this event
+happens, then call this function") and *then* start the loop and hope you
+hooked everything up correctly!  Indeed, much of the ``napari`` source code is
+dedicated to creating and handling events: search the codebase for "``.emit(``"
+and "``.connect(``" to find examples of creating and handling internal events,
+respectively.
+
+If you would like to setup a custom event listener then you  need to hook into
+the napari event.  We offer a couple of convenience decorators to easily
+connect functions to key and mouse events.
+
+Listening for keypress events
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+One option is to use keybindings, that will listen for keypresses and then call
+some callback whenever pressed, with the viewer instance passed as an argument
+to that function. As a basic example, to add a random image to the viewer
+every time the ``i`` key is pressed, and delete the last layer when the ``k``
+key is pressed:
+
+.. code-block:: python
+
+    import numpy as np
+    import napari
+
+    with napari.gui_qt():
+        viewer = napari.Viewer()
+
+        @viewer.bind_key('i')
+        def add_layer(viewer):
+            viewer.add_image(np.random.random((512, 512)))
+
+        @viewer.bind_key('k')
+        def delete_layer(viewer):
+            try:
+                viewer.layers.pop(0)
+            except IndexError:
+                pass
+
+See also this `custom key bindings example
+<https://github.com/napari/napari/blob/master/examples/custom_key_bindings.py>`_
+
+Listening for mouse events
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can also listen for and react to mouse events, like a click or drag event,
+as show here where we update the image with random data every time it is
+clicked.
+
+.. code-block:: python
+
+    import numpy as np
+    import napari
+
+    with napari.gui_qt():
+        viewer = napari.Viewer()
+        layer = viewer.add_image(np.random.random((512, 512)))
+
+        @layer.mouse_drag_callbacks.append
+        def update_layer(layer, event):
+            layer.data = np.random.random((512, 512))
+
+See also the `custom mouse functions
+<https://github.com/napari/napari/blob/master/examples/custom_mouse_functions.py>`_
+and `mouse drag callback
+<https://github.com/napari/napari/blob/master/examples/mouse_drag_callback.py>`_
+examples
+
```

Comment:
It would be good to have a short section here with a list of all the available decorators for mouse/keyboard interactions.